File: lib/features/home/AdFeed.java
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:projects/features/home/ad_card.dart';
import 'package:projects/data/models/ad_model.dart';

class AdFeed extends StatefulWidget {
  final int? categoryId;
  final ScrollController? scrollController;

  const AdFeed({
    super.key,
    this.categoryId,
    this.scrollController,
  });

  @override
  State<AdFeed> createState() => _AdFeedState();
}

class _AdFeedState extends State<AdFeed> {
  List<AdModel> _ads = [];
  bool _isLoading = true;
  final _refreshKey = GlobalKey<RefreshIndicatorState>();

  @override
  void initState() {
    super.initState();
    _loadAds();
  }

  @override
  void didUpdateWidget(covariant AdFeed oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.categoryId != oldWidget.categoryId) {
      _loadAds();
    }
  }

  Future<void> _loadAds() async {
    setState(() => _isLoading = true);

    final cityId = 1;
    final catId = widget.categoryId ?? 0;

    final url = Uri.parse('http://5.59.233.32:8080/ads/public-city/$cityId/category/$catId');
    debugPrint('Загружаю: $url');

    try {
      final response = await http.get(url).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final List<dynamic> jsonData = json.decode(response.body);
        setState(() {
          _ads = jsonData.map((json) {
            return AdModel(
              id: json['id'],
              title: json['category'] ?? '',
              description: json['description'] ?? '',
              phone: json['contact_phone'] ?? '',
              imageUrls: List<String>.from(json['images'] ?? []),
              createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
              categoryId: catId,
            );
          }).toList();
          _isLoading = false;
        });
      } else {
        debugPrint('Ошибка сервера: ${response.statusCode}');
        setState(() => _isLoading = false);
      }
    } catch (e) {
      debugPrint('‼ Ошибка при загрузке: $e');
      setState(() => _isLoading = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final primary = Theme.of(context).primaryColor;

    if (_isLoading) {
      return const Expanded(
        child: Center(child: CircularProgressIndicator()),
      );
    }

    return Expanded(
      child: RefreshIndicator(
        key: _refreshKey,
        onRefresh: _loadAds,
        color: primary,
        child: _ads.isEmpty
            ? ListView(
                controller: widget.scrollController,
                children: const [
                  SizedBox(height: 200),
                  Center(child: Text('Ничего не найдено')),
                ],
              )
            : ListView.builder(
                controller: widget.scrollController,
                padding: const EdgeInsets.symmetric(horizontal: 16),
                itemCount: _ads.length,
                itemBuilder: (_, i) => AdCard(ad: _ads[i]),
              ),
      ),
    );
  }
}

File: lib/features/home/ad_feed.dart
import 'package:flutter/material.dart';
import 'package:easy_localization/easy_localization.dart';
import 'package:nookat996/data/models/ad_model.dart';
import 'package:nookat996/data/services/ad_service.dart';
import 'package:nookat996/features/home/ad_card.dart';
import 'package:nookat996/features/home/ad_card_shimmer.dart';
import 'package:nookat996/features/home/news_header_card.dart';
import 'package:audioplayers/audioplayers.dart';
import '../../utils/sound_helper.dart';

class AdFeed extends StatefulWidget {
  final int? categoryId;
  final ScrollController? scrollController;
  final List<AdModel>? externalAds;

  const AdFeed({
    Key? key,
    this.categoryId,
    this.scrollController,
    this.externalAds,
  }) : super(key: key);

  @override
  State<AdFeed> createState() => _AdFeedState();
}

class _AdFeedState extends State<AdFeed> with SingleTickerProviderStateMixin {
  List<AdModel> ads = [];
  bool isLoading = true;
  bool isLoadingMore = false;
  bool hasMorePages = true;
  int currentPage = 1;
  static const int pageSize = 20;

  final AdService _adService = AdService();
  final ScrollController _scrollController = ScrollController();
  late AnimationController _loadingController;
  late Animation<double> _loadingAnimation;
  bool _isRefreshing = false;

  final GlobalKey<AnimatedListState> _listKey = GlobalKey<AnimatedListState>();

  @override
  void initState() {
    super.initState();
    if (widget.externalAds == null) {
      fetchAds();
    }
    _scrollController.addListener(_onScroll);

    _loadingController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 1500),
    );

    _loadingAnimation = Tween<double>(
      begin: 0.0,
      end: 1.0,
    ).animate(CurvedAnimation(
      parent: _loadingController,
      curve: Curves.easeInOut,
    ));

    _loadingController.repeat();
  }

  @override
  void dispose() {
    _scrollController.removeListener(_onScroll);
    _scrollController.dispose();
    _loadingController.dispose();
    super.dispose();
  }

  void _onScroll() {
    if (_scrollController.position.pixels >=
        _scrollController.position.maxScrollExtent - 200) {
      if (!isLoadingMore && hasMorePages && widget.externalAds == null) {
        loadMoreAds();
      }
    }
  }

  Future<void> fetchAds() async {
    setState(() => isLoading = true);
    currentPage = 1;

    try {
      final response = await _adService.fetchAds(
        categoryId: widget.categoryId,
        page: currentPage,
        pageSize: pageSize,
      );

      if (mounted) {
        setState(() {
          ads = [];
          hasMorePages = response['totalPages'] > currentPage;
          isLoading = false;
        });

        final newAds = response['ads'] as List<AdModel>;
        final bool isNewsCategory = widget.categoryId == 9;

        for (int i = 0; i < newAds.length; i++) {
          ads.add(newAds[i]);
          final animationIndex = isNewsCategory ? i + 1 : i;
          _listKey.currentState?.insertItem(animationIndex,
              duration: const Duration(milliseconds: 350));
        }
      }
    } catch (e) {
      if (mounted) {
        setState(() {
          isLoading = false;
        });
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: Text('error_loading_ads'.tr()),
            duration: const Duration(seconds: 2),
          ),
        );
      }
    }
  }

  Future<void> loadMoreAds() async {
    if (isLoadingMore) return;

    setState(() => isLoadingMore = true);
    currentPage++;

    try {
      final response = await _adService.fetchAds(
        categoryId: widget.categoryId,
        page: currentPage,
        pageSize: pageSize,
      );

      if (mounted) {
        final newAds = response['ads'] as List<AdModel>;
        if (newAds.isEmpty) {
          hasMorePages = false;
        } else {
          final startIndex = ads.length;
          setState(() {
            ads.addAll(newAds);
            hasMorePages = response['totalPages'] > currentPage;
          });

          final bool isNewsCategory = widget.categoryId == 9;
          for (int i = 0; i < newAds.length; i++) {
            final animationIndex =
                isNewsCategory ? startIndex + i + 1 : startIndex + i;
            _listKey.currentState?.insertItem(animationIndex,
                duration: const Duration(milliseconds: 350));
          }
        }
      }
    } catch (e) {
      debugPrint('Ошибка при загрузке дополнительных объявлений: $e');
    } finally {
      if (mounted) {
        setState(() => isLoadingMore = false);
      }
    }
  }

  Future<void> _handleRefresh() async {
    await SoundHelper.playIfEnabled('sounds/refresh.mp3');
    setState(() => _isRefreshing = true);
    await fetchAds();
    setState(() => _isRefreshing = false);
  }

  @override
  Widget build(BuildContext context) {
    final List<AdModel> displayAds = widget.externalAds ?? ads;
    final bool isNewsCategory = widget.categoryId == 9;

    if (isLoading && widget.externalAds == null) {
      return ListView.builder(
        padding: const EdgeInsets.only(top: 11, left: 12, right: 12),
        itemBuilder: (context, index) => const AdCardShimmer(),
      );
    }

    if (displayAds.isEmpty && !isNewsCategory) {
      return Center(child: Text('no_ads_found'.tr()));
    }

    return RefreshIndicator(
      onRefresh: _handleRefresh,
      child: AnimatedList(
        key: _listKey,
        padding: const EdgeInsets.only(left: 12, right: 12, top: 10),
        controller: widget.scrollController ?? _scrollController,
        initialItemCount: displayAds.length +
            (hasMorePages ? 1 : 0) +
            (isNewsCategory ? 1 : 0),
        itemBuilder: (context, index, animation) {
          if (isNewsCategory && index == 0) {
            return const NewsHeaderCard();
          }

          final adjustedIndex = isNewsCategory ? index - 1 : index;

          if (adjustedIndex == displayAds.length) {
            if (isLoadingMore) {
              return Padding(
                padding: const EdgeInsets.symmetric(vertical: 24),
                child: Center(
                  child: CircularProgressIndicator(
                    strokeWidth: 2.5,
                    valueColor: AlwaysStoppedAnimation<Color>(
                      Theme.of(context).primaryColor.withOpacity(0.7),
                    ),
                  ),
                ),
              );
            } else if (hasMorePages) {
              return Padding(
                padding: const EdgeInsets.symmetric(vertical: 16),
                child: Center(
                  child: TextButton(
                    onPressed: loadMoreAds,
                    child: Text(
                      'load_more'.tr(),
                      style: TextStyle(
                        color: Theme.of(context).primaryColor,
                        fontSize: 16,
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                  ),
                ),
              );
            } else {
              return const SizedBox(height: 32);
            }
          }

          return SizeTransition(
            sizeFactor: animation,
            axisAlignment: 0.0,
            child: Padding(
              padding: const EdgeInsets.only(bottom: 10),
              child: AdCard(ad: displayAds[adjustedIndex]),
            ),
          );
        },
      ),
    );
  }
}

// margin: const EdgeInsets.only(bottom: 16),
